File: core/loss.cpp
#include "loss.h"
#include "device.h"
#include <cmath>

TensorPtr MSELoss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->size() == target->size());

    auto diff = prediction->sub(target);
    auto sq = diff->mul(diff);
    return sq->mean();
}

TensorPtr L1Loss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->size() == target->size());

    auto diff = prediction->sub(target);
    auto abs_diff = diff->abs();
    return abs_diff->mean();
}

TensorPtr SmoothL1Loss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->size() == target->size());

    size_t n = prediction->size();
    auto result = Tensor::create({1}, prediction->requires_grad);
    result->data()[0] = 0.0f;

    for (size_t i = 0; i < n; i++) {
        float diff = prediction->data()[i] - target->data()[i];
        float abs_diff = std::fabs(diff);
        if (abs_diff < beta) {
            result->data()[0] += 0.5f * diff * diff / beta;
        } else {
            result->data()[0] += abs_diff - 0.5f * beta;
        }
    }
    result->data()[0] /= static_cast<float>(n);

    if (result->requires_grad) {
        auto pred_ptr = prediction;
        auto target_ptr = target;
        float beta_val = beta;
        result->parents = {pred_ptr};
        result->grad_fn = [pred_ptr, target_ptr, result, n, beta_val]() {
            float scale = result->grad()[0] / static_cast<float>(n);
            for (size_t i = 0; i < n; i++) {
                float diff = pred_ptr->data()[i] - target_ptr->data()[i];
                float abs_diff = std::fabs(diff);
                if (abs_diff < beta_val) {
                    pred_ptr->grad()[i] += scale * diff / beta_val;
                } else {
                    pred_ptr->grad()[i] += scale * (diff > 0 ? 1.0f : -1.0f);
                }
            }
        };
    }

    return result;
}

TensorPtr CrossEntropyLoss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->shape.size() == 2);
    assert(target->shape.size() == 1 || (target->shape.size() == 2 && target->shape[1] == 1));

    // CPU fallback for non-CPU tensors
    if (prediction->device != whitematter::DeviceType::CPU) {
        auto cpu_pred = prediction->to(whitematter::DeviceType::CPU);
        auto cpu_tgt = target->to(whitematter::DeviceType::CPU);
        auto cpu_result = forward(cpu_pred, cpu_tgt);
        cpu_result->to_inplace(prediction->device);
        return cpu_result;
    }

    size_t batch_size = prediction->shape[0];
    size_t num_classes = prediction->shape[1];

    auto log_probs = prediction->log_softmax(-1);

    auto result = Tensor::create({1}, prediction->requires_grad);
    result->data()[0] = 0.0f;

    for (size_t i = 0; i < batch_size; i++) {
        size_t label = static_cast<size_t>(target->data()[i]);
        result->data()[0] -= log_probs->data()[i * num_classes + label];
    }
    result->data()[0] /= static_cast<float>(batch_size);

    if (result->requires_grad) {
        result->parents = {log_probs};
        result->grad_fn = [log_probs, target, result, batch_size, num_classes]() {
            float scale = result->grad()[0] / static_cast<float>(batch_size);
            for (size_t i = 0; i < batch_size; i++) {
                size_t label = static_cast<size_t>(target->data()[i]);
                log_probs->grad()[i * num_classes + label] -= scale;
            }
        };
    }

    return result;
}

TensorPtr NLLLoss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->shape.size() == 2);
    assert(target->shape.size() == 1 || (target->shape.size() == 2 && target->shape[1] == 1));

    size_t batch_size = prediction->shape[0];
    size_t num_classes = prediction->shape[1];

    auto result = Tensor::create({1}, prediction->requires_grad);
    result->data()[0] = 0.0f;

    for (size_t i = 0; i < batch_size; i++) {
        size_t label = static_cast<size_t>(target->data()[i]);
        result->data()[0] -= prediction->data()[i * num_classes + label];
    }
    result->data()[0] /= static_cast<float>(batch_size);

    if (result->requires_grad) {
        auto pred_ptr = prediction;
        result->parents = {pred_ptr};
        result->grad_fn = [pred_ptr, target, result, batch_size, num_classes]() {
            float scale = result->grad()[0] / static_cast<float>(batch_size);
            for (size_t i = 0; i < batch_size; i++) {
                size_t label = static_cast<size_t>(target->data()[i]);
                pred_ptr->grad()[i * num_classes + label] -= scale;
            }
        };
    }

    return result;
}

TensorPtr BCELoss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->size() == target->size());

    size_t n = prediction->size();
    float eps = 1e-7f;

    auto result = Tensor::create({1}, prediction->requires_grad);
    result->data()[0] = 0.0f;

    for (size_t i = 0; i < n; i++) {
        float p = std::max(std::min(prediction->data()[i], 1.0f - eps), eps);
        float y = target->data()[i];
        result->data()[0] -= y * std::log(p) + (1.0f - y) * std::log(1.0f - p);
    }
    result->data()[0] /= static_cast<float>(n);

    if (result->requires_grad) {
        auto pred_ptr = prediction;
        auto target_ptr = target;
        result->parents = {pred_ptr};
        result->grad_fn = [pred_ptr, target_ptr, result, n, eps]() {
            float scale = result->grad()[0] / static_cast<float>(n);
            for (size_t i = 0; i < n; i++) {
                float p = std::max(std::min(pred_ptr->data()[i], 1.0f - eps), eps);
                float y = target_ptr->data()[i];
                pred_ptr->grad()[i] += scale * (-y / p + (1.0f - y) / (1.0f - p));
            }
        };
    }

    return result;
}

TensorPtr BCEWithLogitsLoss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->size() == target->size());

    size_t n = prediction->size();

    auto result = Tensor::create({1}, prediction->requires_grad);
    result->data()[0] = 0.0f;

    for (size_t i = 0; i < n; i++) {
        float x = prediction->data()[i];
        float y = target->data()[i];
        float max_val = std::max(x, 0.0f);
        result->data()[0] += max_val - x * y + std::log(1.0f + std::exp(-std::abs(x)));
    }
    result->data()[0] /= static_cast<float>(n);

    if (result->requires_grad) {
        auto pred_ptr = prediction;
        auto target_ptr = target;
        result->parents = {pred_ptr};
        result->grad_fn = [pred_ptr, target_ptr, result, n]() {
            float scale = result->grad()[0] / static_cast<float>(n);
            for (size_t i = 0; i < n; i++) {
                float x = pred_ptr->data()[i];
                float y = target_ptr->data()[i];
                float sigmoid_x = 1.0f / (1.0f + std::exp(-x));
                pred_ptr->grad()[i] += scale * (sigmoid_x - y);
            }
        };
    }

    return result;
}

TensorPtr KLDivLoss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->size() == target->size());
    assert(prediction->shape.size() >= 1);

    size_t n = prediction->size();
    size_t batch_size = prediction->shape[0];

    auto result = Tensor::create({1}, prediction->requires_grad);
    result->data()[0] = 0.0f;

    for (size_t i = 0; i < n; i++) {
        float t = target->data()[i];
        float log_p = prediction->data()[i];

        if (log_target) {
            float log_t = t;
            float t_prob = std::exp(log_t);
            if (t_prob > 0) {
                result->data()[0] += t_prob * (log_t - log_p);
            }
        } else {
            if (t > 0) {
                result->data()[0] += t * (std::log(t) - log_p);
            }
        }
    }
    result->data()[0] /= static_cast<float>(batch_size);

    if (result->requires_grad) {
        auto pred_ptr = prediction;
        auto target_ptr = target;
        bool log_t = log_target;
        result->parents = {pred_ptr};
        result->grad_fn = [pred_ptr, target_ptr, result, n, batch_size, log_t]() {
            float scale = result->grad()[0] / static_cast<float>(batch_size);
            for (size_t i = 0; i < n; i++) {
                float t = target_ptr->data()[i];
                if (log_t) {
                    pred_ptr->grad()[i] += scale * (-std::exp(t));
                } else {
                    pred_ptr->grad()[i] += scale * (-t);
                }
            }
        };
    }

    return result;
}

TensorPtr FocalLoss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->shape.size() == 2);
    assert(target->shape.size() == 1 || (target->shape.size() == 2 && target->shape[1] == 1));

    size_t batch_size = prediction->shape[0];
    size_t num_classes = prediction->shape[1];

    auto probs = prediction->softmax(-1);

    auto result = Tensor::create({1}, prediction->requires_grad);
    result->data()[0] = 0.0f;

    float eps = 1e-7f;
    for (size_t i = 0; i < batch_size; i++) {
        size_t label = static_cast<size_t>(target->data()[i]);
        float p_t = std::max(probs->data()[i * num_classes + label], eps);
        float focal_weight = std::pow(1.0f - p_t, gamma);
        float loss_i = -focal_weight * std::log(p_t);
        if (alpha >= 0) {
            loss_i *= alpha;
        }
        result->data()[0] += loss_i;
    }
    result->data()[0] /= static_cast<float>(batch_size);

    if (result->requires_grad) {
        auto pred_ptr = prediction;
        auto probs_ptr = probs;
        auto target_ptr = target;
        float gamma_val = gamma;
        float alpha_val = alpha;
        result->parents = {pred_ptr};
        result->grad_fn = [pred_ptr, probs_ptr, target_ptr, result, batch_size, num_classes, gamma_val, alpha_val, eps]() {
            float scale = result->grad()[0] / static_cast<float>(batch_size);
            for (size_t i = 0; i < batch_size; i++) {
                size_t label = static_cast<size_t>(target_ptr->data()[i]);
                float p_t = std::max(probs_ptr->data()[i * num_classes + label], eps);
                float one_minus_pt = 1.0f - p_t;

                for (size_t c = 0; c < num_classes; c++) {
                    float p_c = probs_ptr->data()[i * num_classes + c];
                    float grad_val;

                    if (c == label) {
                        float term1 = gamma_val * std::log(p_t) * std::pow(one_minus_pt, gamma_val);
                        float term2 = -std::pow(one_minus_pt, gamma_val);
                        grad_val = p_c * (term1 + term2) + std::pow(one_minus_pt, gamma_val);
                    } else {
                        float focal_term = std::pow(one_minus_pt, gamma_val - 1.0f);
                        grad_val = -p_c * (focal_term * (gamma_val * p_t * std::log(p_t) + one_minus_pt));
                    }

                    if (alpha_val >= 0) {
                        grad_val *= alpha_val;
                    }
                    pred_ptr->grad()[i * num_classes + c] += scale * grad_val;
                }
            }
        };
    }

    return result;
}

TensorPtr BinaryFocalLoss::forward(const TensorPtr& prediction, const TensorPtr& target) {
    assert(prediction->size() == target->size());

    size_t n = prediction->size();
    float eps = 1e-7f;

    auto result = Tensor::create({1}, prediction->requires_grad);
    result->data()[0] = 0.0f;

    for (size_t i = 0; i < n; i++) {
        float x = prediction->data()[i];
        float y = target->data()[i];

        float p = 1.0f / (1.0f + std::exp(-x));
        p = std::max(std::min(p, 1.0f - eps), eps);

        float p_t = y * p + (1.0f - y) * (1.0f - p);
        float focal_weight = std::pow(1.0f - p_t, gamma);

        float alpha_t = 1.0f;
        if (alpha >= 0) {
            alpha_t = y * alpha + (1.0f - y) * (1.0f - alpha);
        }

        result->data()[0] -= alpha_t * focal_weight * std::log(p_t);
    }
    result->data()[0] /= static_cast<float>(n);

    if (result->requires_grad) {
        auto pred_ptr = prediction;
        auto target_ptr = target;
        float gamma_val = gamma;
        float alpha_val = alpha;
        result->parents = {pred_ptr};
        result->grad_fn = [pred_ptr, target_ptr, result, n, gamma_val, alpha_val, eps]() {
            float scale = result->grad()[0] / static_cast<float>(n);
            for (size_t i = 0; i < n; i++) {
                float x = pred_ptr->data()[i];
                float y = target_ptr->data()[i];

                float p = 1.0f / (1.0f + std::exp(-x));
                p = std::max(std::min(p, 1.0f - eps), eps);

                float p_t = y * p + (1.0f - y) * (1.0f - p);
                float one_minus_pt = 1.0f - p_t;

                float alpha_t = 1.0f;
                if (alpha_val >= 0) {
                    alpha_t = y * alpha_val + (1.0f - y) * (1.0f - alpha_val);
                }

                float dp_dx = p * (1.0f - p);
                float dp_t_dx = (2.0f * y - 1.0f) * dp_dx;

                float focal_grad;
                if (gamma_val == 0) {
                    focal_grad = -alpha_t / p_t * dp_t_dx;
                } else {
                    float log_pt = std::log(p_t);
                    float bracket = one_minus_pt / p_t - gamma_val * log_pt;
                    focal_grad = -alpha_t * dp_t_dx * std::pow(one_minus_pt, gamma_val - 1.0f) * bracket;
                }

                pred_ptr->grad()[i] += scale * focal_grad;
            }
        };
    }

    return result;
}

File: whitematter/nn/loss.py
"""Loss functions."""

from __future__ import annotations

import numpy as np
from ..tensor import Tensor, _acc
from .module import Module


class MSELoss(Module):
    """Mean Squared Error loss."""

    def forward(self, pred: Tensor, target: Tensor) -> Tensor:
        diff = pred.sub(target)
        return diff.mul(diff).mean()


class L1Loss(Module):
    """Mean Absolute Error loss."""

    def forward(self, pred: Tensor, target: Tensor) -> Tensor:
        return pred.sub(target).abs().mean()


class SmoothL1Loss(Module):
    """Smooth L1 (Huber) loss."""

    def __init__(self, beta: float = 1.0):
        super().__init__()
        self.beta = beta

    def forward(self, pred: Tensor, target: Tensor) -> Tensor:
        diff = pred.sub(target)
        abs_diff = diff.abs()
        beta = self.beta
        # Piecewise: 0.5*x^2/beta if |x|<beta, else |x|-0.5*beta
        d = diff.data
        ad = np.abs(d)
        loss_data = np.where(ad < beta, 0.5 * d ** 2 / beta, ad - 0.5 * beta)
        result = pred._make_result(np.atleast_1d(loss_data.mean()), [pred, target])
        if result.requires_grad:
            n = float(pred.data.size)
            def _backward(grad):
                scale = grad.flat[0] / n
                g = np.where(ad < beta, d / beta, np.sign(d)) * scale
                if pred.requires_grad:
                    pred.grad = _acc(pred.grad, g.astype(np.float32))
                if target.requires_grad:
                    target.grad = _acc(target.grad, (-g).astype(np.float32))
            result._grad_fn = _backward
        return result


class CrossEntropyLoss(Module):
    """Cross-entropy loss (expects raw logits and integer labels)."""

    def forward(self, logits: Tensor, targets: Tensor) -> Tensor:
        # logits: (N, C), targets: (N,) integer labels
        x = logits.data
        t = targets.data.astype(np.int64)
        N = x.shape[0]
        # Numerically stable log-softmax
        x_max = x.max(axis=1, keepdims=True)
        log_sum_exp = x_max + np.log(np.exp(x - x_max).sum(axis=1, keepdims=True))
        log_probs = x - log_sum_exp
        loss_val = -log_probs[np.arange(N), t].mean()

        result = logits._make_result(np.atleast_1d(loss_val), [logits])
        if result.requires_grad:
            probs = np.exp(log_probs)
            def _backward(grad):
                if logits.requires_grad:
                    g = probs.copy()
                    g[np.arange(N), t] -= 1.0
                    g *= grad.flat[0] / N
                    logits.grad = _acc(logits.grad, g.astype(np.float32))
            result._grad_fn = _backward
        return result


class NLLLoss(Module):
    """Negative log-likelihood loss (expects log-probabilities)."""

    def forward(self, log_probs: Tensor, targets: Tensor) -> Tensor:
        lp = log_probs.data
        t = targets.data.astype(np.int64)
        N = lp.shape[0]
        loss_val = -lp[np.arange(N), t].mean()

        result = log_probs._make_result(np.atleast_1d(loss_val), [log_probs])
        if result.requires_grad:
            def _backward(grad):
                if log_probs.requires_grad:
                    g = np.zeros_like(lp)
                    g[np.arange(N), t] = -1.0 / N
                    log_probs.grad = _acc(log_probs.grad, (g * grad.flat[0]).astype(np.float32))
            result._grad_fn = _backward
        return result


class BCELoss(Module):
    """Binary cross-entropy loss (expects probabilities)."""

    def forward(self, pred: Tensor, target: Tensor) -> Tensor:
        p = np.clip(pred.data, 1e-7, 1 - 1e-7)
        y = target.data
        loss = -(y * np.log(p) + (1 - y) * np.log(1 - p))
        result = pred._make_result(np.atleast_1d(loss.mean()), [pred, target])
        if result.requires_grad:
            n = float(pred.data.size)
            def _backward(grad):
                scale = grad.flat[0] / n
                g = (-y / p + (1 - y) / (1 - p)) * scale
                if pred.requires_grad:
                    pred.grad = _acc(pred.grad, g.astype(np.float32))
            result._grad_fn = _backward
        return result


class BCEWithLogitsLoss(Module):
    """Binary cross-entropy with logits (numerically stable)."""

    def forward(self, logits: Tensor, target: Tensor) -> Tensor:
        x = logits.data
        y = target.data
        # max(x,0) - x*y + log(1+exp(-|x|))
        loss = np.maximum(x, 0) - x * y + np.log1p(np.exp(-np.abs(x)))
        result = logits._make_result(np.atleast_1d(loss.mean()), [logits, target])
        if result.requires_grad:
            sig = 1.0 / (1.0 + np.exp(-np.clip(x, -88, 88)))
            n = float(logits.data.size)
            def _backward(grad):
                scale = grad.flat[0] / n
                g = (sig - y) * scale
                if logits.requires_grad:
                    logits.grad = _acc(logits.grad, g.astype(np.float32))
            result._grad_fn = _backward
        return result


class KLDivLoss(Module):
    """KL divergence loss."""

    def __init__(self, log_target: bool = False):
        super().__init__()
        self.log_target = log_target

    def forward(self, pred: Tensor, target: Tensor) -> Tensor:
        # pred = log(Q), target = P (or log(P))
        p = pred.data
        if self.log_target:
            t_exp = np.exp(target.data)
            loss = (t_exp * (target.data - p)).sum() / p.shape[0]
        else:
            t = target.data
            loss = (t * (np.log(t + 1e-8) - p)).sum() / p.shape[0]

        result = pred._make_result(np.atleast_1d(loss), [pred])
        if result.requires_grad:
            N = p.shape[0]
            def _backward(grad):
                if pred.requires_grad:
                    if self.log_target:
                        g = -np.exp(target.data) / N
                    else:
                        g = -target.data / N
                    pred.grad = _acc(pred.grad, (g * grad.flat[0]).astype(np.float32))
            result._grad_fn = _backward
        return result


class FocalLoss(Module):
    """Focal loss for class-imbalanced classification."""

    def __init__(self, gamma: float = 2.0, alpha: float = 1.0):
        super().__init__()
        self.gamma = gamma
        self.alpha = alpha

    def forward(self, logits: Tensor, targets: Tensor) -> Tensor:
        x = logits.data
        t = targets.data.astype(np.int64)
        N = x.shape[0]
        gamma, alpha = self.gamma, self.alpha

        x_max = x.max(axis=1, keepdims=True)
        e = np.exp(x - x_max)
        probs = e / e.sum(axis=1, keepdims=True)

        p_t = probs[np.arange(N), t]
        focal_weight = (1 - p_t) ** gamma
        loss = -alpha * focal_weight * np.log(p_t + 1e-8)
        result = logits._make_result(np.atleast_1d(loss.mean()), [logits])
        if result.requires_grad:
            def _backward(grad):
                if logits.requires_grad:
                    g2 = probs.copy()
                    for i in range(N):
                        pt_i = p_t[i]
                        log_pt_i = np.log(pt_i + 1e-8)
                        fw = (1 - pt_i) ** gamma
                        dfw = -gamma * (1 - pt_i) ** (gamma - 1)
                        for c in range(x.shape[1]):
                            # dsoftmax/dlogit: p_c * (delta_ct - p_t)
                            ds = probs[i, c] * ((1.0 if c == t[i] else 0.0) - probs[i, t[i]])
                            # d(loss_i)/dlogit_c = -alpha * (dfw * ds * log_pt + fw * ds / pt)
                            g2[i, c] = -alpha * (dfw * ds * log_pt_i + fw * ds / (pt_i + 1e-8))
                    logits.grad = _acc(logits.grad, (g2 * grad.flat[0] / N).astype(np.float32))
            result._grad_fn = _backward
        return result


class BinaryFocalLoss(Module):
    """Binary focal loss with sigmoid."""

    def __init__(self, gamma: float = 2.0, alpha: float = 1.0):
        super().__init__()
        self.gamma = gamma
        self.alpha = alpha

    def forward(self, logits: Tensor, target: Tensor) -> Tensor:
        x = logits.data
        y = target.data
        gamma, alpha = self.gamma, self.alpha

        sig = 1.0 / (1.0 + np.exp(-np.clip(x, -88, 88)))
        p_t = y * sig + (1 - y) * (1 - sig)
        focal_weight = (1 - p_t) ** gamma
        bce = np.maximum(x, 0) - x * y + np.log1p(np.exp(-np.abs(x)))
        loss = alpha * focal_weight * bce

        result = logits._make_result(np.atleast_1d(loss.mean()), [logits])
        if result.requires_grad:
            n = float(logits.data.size)
            def _backward(grad):
                if logits.requires_grad:
                    # d/dx = alpha * [dfw * bce + fw * dbce]
                    dp_t = y * sig * (1 - sig) + (1 - y) * (-sig * (1 - sig))
                    dfw = -gamma * (1 - p_t) ** (gamma - 1) * dp_t
                    dbce = sig - y
                    g = alpha * (dfw * bce + focal_weight * dbce)
                    logits.grad = _acc(logits.grad, (g * grad.flat[0] / n).astype(np.float32))
            result._grad_fn = _backward
        return result
